File: content/LLD/Reference-Code/2. SOLID Principles/3. Liskov Substitution Principle (LSP)/2. Good Example/savings_account.py
from withdrawable_account import WithdrawableAccount


class SavingsAccount(WithdrawableAccount):
    def __init__(self, amount):
        super().__init__(amount)

    def deposit(self, amount):
        self.balance += amount
        print(f"Amount deposited, current balance = {self.balance}")

    def withdraw(self, amount):
        if self.balance < amount:
            print("Cannot withdraw, not enough balance")
        else:
            self.balance -= amount
            print(f"Amount withdrawn, current balance = {self.balance}")

File: content/LLD/Reference-Code/2. SOLID Principles/3. Liskov Substitution Principle (LSP)/2. Good Example/SavingsAccount.java
public class SavingsAccount extends WithdrawableAccount {
    public SavingsAccount(int amount) {
        super(amount);
    }

    @Override
    public void deposit(int amount) {
        this.balance += amount;
        System.out.println("Amount deposited, current balance = " + this.balance);
    }

    @Override
    public void withdraw(int amount) {
        if (this.balance < amount) {
            System.out.println("Cannot withdraw, not enough balance");
        } else {
            this.balance -= amount;
            System.out.println("Amount withdrawn, current balance = " + this.balance);
        }
    }
}
